Automattic/newspack-plugin
# SalesForce integration

### All Submissions:






### Changes proposed in this Pull Request:

Adds a Salesforce config view to the Reader Revenue wizard which allows admin users to connect a Salesforce org via the SF [OAuth 2.0 web server flow](https://help.salesforce.com/articleView?id=remoteaccess_oauth_web_server_flow.htm&type=5).

When the Newspack site is successfully connected to a Salesforce org, a WooCommerce webhook is generated which will trigger a data sync every time a donation order is created. This webhook is deleted if the Salesforce config is reset.

When an order is created, the webhook fires a request which includes order and billing details. This data is then synced to the connected Salesforce org via the Salesforce REST API. Lead records are looked up by email address and if there's an existing Lead record in SF, donor details are added to that record. If there isn't an existing Lead record, one is created. More details on the SalesForce API here: https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/intro_what_is_rest_api.htm

<img width="701" alt="Screen Shot 2020-06-05 at 3 35 18 PM" src="https://user-images.githubusercontent.com/2230142/83924586-35c4ee00-a742-11ea-8f2b-6e3886a075c9.png">

At this stage, this is a proof-of-concept implementation, so I'd like to request feedback on the overall approach as well as the UI and where in the wizard it lives, etc. before I consider it ready for a full review/merge. We may want to split this out into an entirely new wizard (Customer Retention?), but for now, the existing Reader Revenue wizard seemed like the best home for this.

I have not tested this in browsers other than Chrome so far, nor written automated tests for this functionality. So far I've only tested it in a fully setup Newspack site, not as part of a new site setup.



### How to test the changes in this Pull Request:

1. Pull and build this branch to a local Newspack site environment. Run `npm install && npm run dev` to rebuild JS.
2. Visit the Reader Revenue wizard to find the Configure SalesForce page.
3. Sign up for a new SalesForce Developer Org: https://developer.salesforce.com/signup
4. Once your org has been created, follow these instructions to create a new Connected App: https://help.salesforce.com/articleView?id=connected_app_create.htm
  - In short, go to Setup > Apps > App Manager > New Connected App and fill out the form, making sure to enable OAuth settings with the following capabilities: api, refresh_token, offline_access, web
  - Make sure to paste the URL for your Configure SalesForce page into the "Redirect URI" field. This whitelists the URL for use in the OAuth flow, otherwise SalesForce will block access to the login/allow screens in the next few steps.
5. Note the Consumer Key and Consumer Secret strings that are generated in the created Connected App page.
6. In the Configure SalesForce page, enter your Connected App's Consumer Key and Secret into the text fields. The "Connect" button should become available after both fields contain values.
7. After clicking "Connect," you should be redirected to a SalesForce login page (if not already logged in with the same browser), or to an "Allow Access?" page. Click Allow to grant access.
8. You will be redirected back to the wizard page on the Newspack site with an authorization code appended to the URL. The app should automatically fetch API access and refresh tokens using the auth code, then clear the `code` param from the URL.
9. In theory, the site can now use the `access_token` to make requests to the SalesForce org via REST API, or use the `refresh_token` to fetch new `access_token` values.

### Other information:
